lib/fastlane/plugin/firebase_app_distribution/actions/login.rb

```diff
@@ -0,0 +1,65 @@
+require 'googleauth'
+require 'googleauth/stores/file_token_store'
+require "google/apis/people_v1"
+# require "googleauth"
+# require "googleauth/stores/file_token_store"
+require "fileutils"
+module Fastlane
+  module Actions
+    module SharedValues
+      TEST_AUTH_CUSTOM_VALUE = :TEST_AUTH_CUSTOM_VALUE
+    end
+
+    class LoginAction < Action
+      def self.run(params)
+        client_id = Google::Auth::ClientId.new("563584335869-fgrhgmd47bqnekij5i8b5pr03ho849e6.apps.googleusercontent.com", "j9iVZfS8kkCEFUPaAeJV0sAi")
+        scope = 'https://www.googleapis.com/auth/cloud-platform'
+        token_store = nil
+        authorizer = Google::Auth::UserAuthorizer.new(client_id, scope, token_store)
+        url = authorizer.get_authorization_url(base_url: "urn:ietf:wg:oauth:2.0:oob")
+
+        UI.message("Please open the following address in your browser:")
+        UI.message(url.to_s)
+        UI.message(" ")
+        code = UI.input("Enter the resulting code here: ")
+        credentials = authorizer.get_credentials_from_code(code: code, base_url: "urn:ietf:wg:oauth:2.0:oob")
+        token = credentials.refresh_token
+
+        UI.message("Refresh Token: #{token}")
+        UI.message(" ")
+        UI.message("Set the refresh token as a FIREBASE_TOKEN enviorment variable")
+      rescue
```

Comment:
I think it would be good to have two `rescue` clauses here. One that rescues just the specific authentication error that you got when entering an invalid or malformed code. The other would rescue any other error.

The message for the first could be something like:  "The code you entered was invalid. Ensure that you have copied the code correctly."

The message for the second could be more general, i.e.: "An error occurred."

Also, for these messages you should use UI.error, and consider using UI.crash for the second case, since some sort of unexpected error happened in that case. See [this guide]( https://docs.fastlane.tools/advanced/actions/#user-input-and-output) and see [this example](https://github.com/fastlane/fastlane/blob/334d3789c37ad0af236fdd39696d7684bf41bba7/match/lib/match/encryption/openssl.rb#L131) as well.